pulp_python/app/viewsets.py

```diff
@@ -70,18 +70,33 @@ def publish(self, request, pk):
         """
         Dispatches a publish task.
         """
-        try:
-            repository_uri = request.data['repository']
-        except KeyError:
-            raise ValidationError(detail=_('Repository URI must be specified.'))
-
         publisher = self.get_object()
-        repository = self.get_resource(repository_uri, Repository)
+        repository = None
+        repository_version = None
+
+        if 'repository' not in request.data and 'repository_version' not in request.data:
+            raise ValidationError("Either the 'repository' or 'repository_version' "
+                                  "need to be specified.")
+
+        if 'repository' in request.data and request.data['repository']:
+            repository = self.get_resource(request.data['repository'], Repository)
+
+        if 'repository_version' in request.data and request.data['repository_version']:
+            repository_version = self.get_resource(request.data['repository_version'],
+                                                   RepositoryVersion)
+
+        if repository and repository_version:
```

Comment:
I'm not convinced this is necessary. If both are specified, why not just use the more specific of the 2, (version). `repository` is still correct.

Reply:
I was following what was laid out by the last sentence here: https://pulp.plan.io/issues/3360

Reply:
I don't follow, but I can see that there could be some confusion, particularly if the RepositoryVersion.repository doesn't match Respository. Please leave this as you did it.